### PR TITLE
Add all exports for react-native-elements

### PR DESCRIPTION
**Pull request by ifiokjr**
This continues the unmerged work from #17886. 

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: <<url here>>
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

### CODE REVIEW

**Comment by iRoachie** [comment]
Your formatting right now breaks the Travis build. Will review the other info later

**Comment by ifiokjr** [comment]
My VS Code settings over eagerly import modules. These two lines somehow snuck in. 

```ts
import { stringify } from '../wellknown/index';
import { numberTypeAnnotation } from '../babel-types/index';
```

I've removed them now.

File: types/react-native-elements/README.md [hunk 1]
```diff
@@ -6,22 +6,22 @@ List will update as typings are added
 - [x] Buttons
 - [x] Badge
 - [ ] Social Icons / Social Icon Buttons
-- [ ] Icons
-- [ ] Side Menu
-- [ ] Form Elements
+- [x] Icons
+- [ ] ~~Side Menu~~ DEPRECATED
```
Comment (iRoachie) [inline types/react-native-elements/README.md]: Will still need to have typings for deprecated components.

File: types/react-native-elements/index.d.ts [hunk 2]
```diff
@@ -1,11 +1,51 @@
...
+    ViewProperties,
+    TextInputProperties,
+    TextInput,
+    StatusBarProperties,
+    KeyboardType,
+    KeyboardTypeIOS,
+    StyleProp,
+} from 'react-native';
+
+export type IconType =
```
Comment (iRoachie) [inline types/react-native-elements/index.d.ts]: This restricts custom iconSets the user may implement. See https://github.com/oblador/react-native-vector-icons#custom-fonts

File: types/react-native-elements/index.d.ts [hunk 3]
```diff
@@ -250,4 +400,1126 @@ export interface BadgeProperties {
...
+ *
+ * @see https://react-native-training.github.io/react-native-elements/API/card/
+ */
+export class Card extends React.Component<CardProps, any> {}
+
+/**
+ * Set the buttons within a Group.
+ */
+export interface ElementObject {
+    element: JSX.Element | React.ReactType;
```
Comment (iRoachie) [inline types/react-native-elements/index.d.ts]: I think what you wanted here was 
`type ElementObject = JSX.Element | React.ReactType`

File: types/react-native-elements/index.d.ts [hunk 4]
```diff
@@ -250,4 +400,1126 @@ export interface BadgeProperties {
...
+export interface ButtonGroupProps {
+    /**
+     * Current selected index of array of buttons
+     */
+    selectedIndex: number;
+
+    /**
+     * Method to update Button Group Index
+     */
+    // onPress(): void;
```
Comment (iRoachie) [inline types/react-native-elements/index.d.ts]: Remove?

File: types/react-native-elements/index.d.ts [hunk 5]
```diff
@@ -250,4 +400,1126 @@ export interface BadgeProperties {
...
+     */
+    component?: React.ComponentClass;
+
+    /**
+     * Specify styling for main button container
+     */
+    containerStyle?: ViewStyle;
+
+    /**
+     * inherited styling	object (style)	specify styling for button
```
Comment (iRoachie) [inline types/react-native-elements/index.d.ts]: Spacing

File: types/react-native-elements/index.d.ts [hunk 6]
```diff
@@ -250,4 +400,1126 @@ export interface BadgeProperties {
...
+     */
+    buttonStyle?: ViewStyle;
```
Comment (iRoachie) [inline types/react-native-elements/index.d.ts]: All the instances of `ViewStyle, TextStyle, ImageStyle` need to be wrapped with the `StyleProp` from 'react-native`

e.g `StyleProp<ViewStyle>`. This is so someone can pass in an array of styles

File: types/react-native-elements/index.d.ts [hunk 7]
```diff
@@ -250,4 +400,1126 @@ export interface BadgeProperties {
...
+
+    /**
+     * Determines what the opacity of the wrapped view should be when touch is active.
+     */
+    activeOpacity?: number;
+
+    /**
+     * Border radius for the container
+     *
+     * @type {number}
```
Comment (iRoachie) [inline types/react-native-elements/index.d.ts]: Comments like these you can remove "type, memberof, export" they just add bloat.

File: types/react-native-elements/index.d.ts [hunk 8]
```diff
@@ -250,4 +400,1126 @@ export interface BadgeProperties {
...
+    /**
+     * onPress method
+     */
+    onPress?(): void;
+
+    children: React.ReactNode;
+}
+
+/**
+ * DEPRECATED
```
Comment (iRoachie) [inline types/react-native-elements/index.d.ts]: You can remove this add just add the @deprecated tag

**Review by iRoachie — CHANGES_REQUESTED** [review]
Fantastic work mate. Just fix up the issues and let's get this merged

**Comment by ifiokjr** [comment]
@iRoachie 

I've added all relevant types and it's now working for my codebase. It's currently good enough for my usage, although there are a few issues that will crop up. 

- IconObject covers too many scenarios. In RNE, each component which needs an Icon defines a slightly different interface for it. Rather than defining marginally different interfaces for each component I've left the different attributes all on the main IconObject. This should probably change. 
```ts
export interface IconObject {
    name?: string;
    color?: string;
    size?: number;
    type?: IconType;
    style?: StyleProp<ViewStyle>;
}
```

- `react-native-vector-icons` doesn't export its Icon type and hence, I've left `getIconType` with an `any` return type. 


Feel free to point out other problems.

File: types/react-native-elements/index.d.ts (hunk 3 above)
Comment (ifiokjr) [reply types/react-native-elements/index.d.ts]: This is actually intentional. It's used in the `ButtonGroupProps` - buttons property and the `ListItemProps` - badge property.

**Comment by iRoachie** [comment]
Smart by making them optional ha! I would've done the same. Looking over now

File: types/react-native-elements/index.d.ts [hunk 9]
```diff
@@ -1,11 +1,60 @@
-// Type definitions for react-native-elements 0.13
+// Type definitions for react-native-elements 0.16
 // Project: https://github.com/react-native-training/react-native-elements#readme
 // Definitions by: Kyle Roach <https://github.com/iRoachie>
+//                 Ifiok Jr. <https://github.com/ifiokjr>
```
Comment (iRoachie) [inline types/react-native-elements/index.d.ts]: Adding your name here means you become a maintainer of the project which means you'll automatically get email/notified of all issues and PRs for this lib. Just making sure you know what it means

File: types/react-native-elements/README.md [hunk 10]
```diff
@@ -5,23 +5,23 @@ List will update as typings are added
...
-- [ ] Form Elements
-- [ ] Search Bar
-- [ ] ButtonGroup
-- [ ] Checkboxes
-- [ ] List Element
-- [ ] Badge
-- [ ] Tab Bar Component
+- [x] Social Icons / Social Icon Buttons
+- [x] Icons
+- [x] ~~Side Menu~~ DEPRECATED
```
Comment (iRoachie) [inline types/react-native-elements/README.md]: Can remove the ~ and DEPRECATED.

**Review by iRoachie — CHANGES_REQUESTED** [review]
Great work again. The tests are passing. Let's get this merged and out in the wild. If anything we can circle back and make changes.

File: types/react-native-elements/index.d.ts (hunk 9 above)
Comment (ifiokjr) [reply types/react-native-elements/index.d.ts]: That's okay with me.

**Review by iRoachie — APPROVED** [review]

**Comment by ifiokjr** [comment]
Any ideas on when this can be merged? I'm hoping to start using the definitions in a couple of my projects soon.

**Comment by iRoachie** [comment]
@alloy Could you check this?

**Comment by iRoachie** [comment]
Ha, thanks again

**Pull request merged**